mcproto/types/bitset.py

```diff
@@ -0,0 +1,194 @@
+    def from_int(cls, n: int) -> FixedBitset:
+        """Return a new FixedBitset with the given integer value.
+
+        :param n: The integer value.
+        """
+        if cls.__n == -1:
+            raise ValueError("Bitset size is not defined.")
+        if n < 0:
+            # Manually compute two's complement
+            n = -n
+            data = bytearray(n.to_bytes(math.ceil(cls.__n / 8), "big"))
+            for i in range(len(data)):
+                data[i] ^= 0xFF
+            data[-1] += 1
+        else:
+            data = bytearray(n.to_bytes(math.ceil(cls.__n / 8), "big"))
+        return cls(data=data)
+
+    def __setitem__(self, index: int, value: bool) -> None:
+        byte_index = index // 8
+        bit_index = index % 8
+        if value:
+            self.data[byte_index] |= 1 << bit_index
+        else:
+            self.data[byte_index] &= ~(1 << bit_index)
+
+    def __getitem__(self, index: int) -> bool:
+        byte_index = index // 8
+        bit_index = index % 8
+        return bool(self.data[byte_index] & (1 << bit_index))
+
+    def __len__(self) -> int:
+        return self.__n
+
+    def __and__(self, other: FixedBitset) -> FixedBitset:
+        if self.__n != other.__n:
+            raise ValueError("Bitsets must have the same size.")
+        return type(self)(data=bytearray(a & b for a, b in zip(self.data, other.data)))
+
+    def __or__(self, other: FixedBitset) -> FixedBitset:
+        if self.__n != other.__n:
+            raise ValueError("Bitsets must have the same size.")
+        return type(self)(data=bytearray(a | b for a, b in zip(self.data, other.data)))
+
+    def __xor__(self, other: FixedBitset) -> FixedBitset:
+        if self.__n != other.__n:
+            raise ValueError("Bitsets must have the same size.")
+        return type(self)(data=bytearray(a ^ b for a, b in zip(self.data, other.data)))
+
+    def __invert__(self) -> FixedBitset:
+        return type(self)(data=bytearray(~a & 0xFF for a in self.data))
+
+    def __bytes__(self) -> bytes:
+        return bytes(self.data)
+
+    @override
+    def __eq__(self, value: object) -> bool:
+        if not isinstance(value, FixedBitset):
+            return NotImplemented
+        return self.data == value.data and self.__n == value.__n
+
+
+@define
+class Bitset(MCType):
+    """Represents a lenght-prefixed bitset with a variable size.
+
+    :param size: The number of longs in the array representing the bitset.
+    :param data: The bits of the bitset.
+    """
+
+    size: int
+    data: list[int]
+
+    @override
+    def serialize_to(self, buf: Buffer) -> None:
+        buf.write_varint(self.size)
+        for i in range(self.size):
+            buf.write_value(StructFormat.LONGLONG, self.data[i])
+
+    @override
+    @classmethod
+    def deserialize(cls, buf: Buffer) -> Bitset:
+        size = buf.read_varint()
+        if buf.remaining < size * 8:
+            raise IOError("Not enough data to read bitset.")
```

Comment:
Not really, but because we want to read a big chunk of data, I didn't want to have an error that only says impossible to read 1 LONGLONG even tho we are reading 32 at once.

On the other hand the `IOError` already says what is happening and the traceback will indicate where it comes from so it's not necessary.